src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,152 @@
+            tail.next = node;
+            tail = node;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index);
+        Node node = new Node<>(null, value, null);
+        if (index == size) {
+            add(value);
+            return;
+        } else if (index == 0) {
+            node.next = head;
+            head.prev = node;
+            head = node;
+        } else {
+            Node<T> current = findNodeByIndex(index);
+            node.next = current;
+            node.prev = current.prev;
+            current.prev.next = node;
+            current.prev = node;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T listElement : list) {
+            add(listElement);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = findNodeByIndex(index);
+        T item = node.item;
+        node.item = value;
+        return item;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> current = findNodeByIndex(index);
+        remover(current);
+        return current.item;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> current = findNodeByObject(object);
+        if (current == null) {
+            return false;
+        }
+        remover(current);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index);
+        }
+        Node<T> currentElement;
+        if (index < size / 2) {
+            currentElement = head;
+            for (int i = 0; i < index; i++) {
+                currentElement = currentElement.next;
+            }
+        } else {
+            currentElement = tail;
+            for (int i = size - 1; i > index; i--) {
+                currentElement = currentElement.prev;
+            }
+        }
+        return currentElement;
+    }
+
+    private Node<T> findNodeByObject(T object) {
+        for (Node<T> currentNode = head; currentNode != null; currentNode = currentNode.next) {
+            if ((currentNode.item == object)
+                    || (currentNode.item != null && currentNode.item.equals(object))) {
+                return currentNode;
+            }
+        }
+        return null;
+    }
+
+    private void remover(Node<T> current) {
+        if (current.prev == null && current.next == null) {
+            head = null;
+        } else if (current.equals(head)) {
```

Comment:
let's use `==` instead of `equals` here, we need to be sure that nodes have the same link. or simple check `current.prev == null`